python/ray/serve/_private/router.py

```diff
@@ -593,16 +619,26 @@ async def choose_two_replicas_with_backoff(
                         )
                     )
             elif (
-                self._prefer_local_routing
-                and backoff_index == 0
-                and len(self._colocated_replica_ids) > 0
+                self._prefer_local_node_routing
+                and not tried_same_node
+                and len(self._colocated_replica_ids[LocalityScope.NODE]) > 0
+            ):
+                # Attempt to schedule requests to replicas on the same node at most once
+                candidate_replica_ids = self._colocated_replica_ids[LocalityScope.NODE]
+                tried_same_node = True
+            elif (
+                not tried_same_az
+                and len(self._colocated_replica_ids[LocalityScope.AVAILABILITY_ZONE])
```

Comment:
I believe this is missing a check for `self._prefer_local_az_routing`, no?

unless I'm missing something, if the tests are passing as-is then they are insufficient

Reply:
Yeah that was missing, added back in.